### DISCUSSION TITLE
Microservices Config Maps - Preveserve file flag behaviour

### DISCUSSION THREAD

**Discussion by Noste92** [mia-platform/community#185] [Q&A] [answered]
Hi,

we have noticed something unexpected with the flag "Preserve files and directories already existing in the Runtime Mount Path directory" configuring a Config Map.

We have a microservices written in Node that requires a JSON file when consuming a specific Kafka message, like this:

```js
const config = require(PATH_TO_CONFIG);
```

With the preserve files flag enabled the service returns a MODULE_NOT_FOUND error, it seems that the file configured is not present in the directory, but it is correctly configured in the config map section.
We have solved this issue by disabling that flag.

Reading the public documentation is not very clear about the behaviour of the flag.
https://docs.mia-platform.eu/docs/development_suite/api-console/api-design/services#add-a-configuration.

If the flag is enabled and I have configured a file under a specific path, is it possible that the file is absent?

Thanks,
Stefano

**Comment by fredmaggiowski** [comment, marked as answer]
Hi @Noste92 

The purpose of the flag is to prevent the mounted configmap to override existing files in the image you are running.

Suppose your image has a `/app` directory with a bunch of files. If you try to mount a ConfigMap in the `/app` directory, the contents of the ConfigMap will entirely replace the contents of the original directory (removing already existing files regardless the contents of the ConfigMap).

With the flag, existing files are preserved and the ConfigMap only mounts the specific files you require. It seems pretty odd that it worked without the flag 🤔 

Can you provide the actual configurations you are using?

**Reply by Noste92** [reply]
Hi @fredmaggiowski,

Thanks for your explanation.

This is the configuration of the config map, I have removed some parts due to privacy:
![image](https://user-images.githubusercontent.com/14860750/216067825-5f1c1aa5-a1db-4dcd-985e-98a3043e7d6f.png)

I have tried to understand how we have encountered this problem, I think I got it.

 We merge two branches:
- one is configured with two files and the flag preserve files set to false
- the other one is configured with only one file and the flag preserve files set to true 

During the merge we probably solved a conflict like this, keeping only a subpath:
![image](https://user-images.githubusercontent.com/14860750/216086499-b062f4f3-d6bd-4633-aed3-de11a4409317.png)

Do you that this could cause a situation like the one described?
I don't think is very clear that this is the result of the flag being enabled.

**Reply by fredmaggiowski** [reply]
Hey, `subPaths` is the configuration beneath the "preserve" checkbox, however I don't understand what you expect now the behaviour to be; if the flag is disabled then any function previously existing in the `/home/node/app/configs` directory will be removed; are you trying to import something else from there other that the json file you are mounting? 

Or is the problem that only one subPath is present while you have two files in the ConfigMap? 🤔

**Reply by Noste92** [reply]
Hi, 
the problem is that only one subPath is present while I have two files in the ConfigMap, and I'm not trying to import files not present in the subPaths configured.
After this merge was not easy to realize from the console that was the turning point.

**Reply by fredmaggiowski** [reply]
ok, sorry but I'm still not getting if you are still affected by the problem or if it was an oversight that you now resolved 😅

**Reply by Noste92** [reply]
Hi @fredmaggiowski,

Sorry, I've messed up a little bit. I will try to resume what happened.

When I opened this discussion, I thought the problem was related to the "preserve" flag so I would like to know how we solved the problem by disabling the flag.

After your answer, by reading the commits I discovered that a bad conflict resolution generated the problem during a merge.

In my opinion, I think that the problem it's solved, we are happy with this configuration, so we can close the discussion. 

I've only noticed that it's not explicit, during the merge,  that modifying the subpaths will affect the preserve files flag. 

Thank you

**Reply by fredmaggiowski** [reply]
thank you for the clarification; I understand the issue now, the merge does not provide any insight over how the configurations you are merging actually works so I get why you got confused :)

I'd mark the answer as accepted then. @albertotessarotto FYI lets keep track of this feedback about the merge feature.
